File: ProblemSolve/905.cpp
/************************
 * Sort Array By Parity *
 ************************/

class Solution {
public:
    vector<int> sortArrayByParity(vector<int>& nums) {
        int left = 0;
        int right = nums.size() - 1;
        
        //배열 양 끝을 확인하면서 홀수를 뒤로 몰아넣는다
        while(left < right){
            //홀수를 뒤쪽으로 몰아넣는다
            if(nums[left] % 2 != 0){
                swap(nums[left], nums[right]);
                right--;
            }
            //짝수인 경우 그대로 둔다
            else{
                left++;
            }
        }
        
        return nums;
    }
};

File: ProblemSolve/905.cs
/************************
 * Sort Array By Parity *
 ************************/

public class Solution {
    public void Swap(ref int a, ref int b){
        int temp = a;
        a = b;
        b = temp;
    }
    
    public int[] SortArrayByParity(int[] nums) {
        int leftIdx = 0;
        int rightIdx = nums.Length - 1;
        
        while(leftIdx < rightIdx){
            if(nums[leftIdx] % 2 != 0){
                Swap(ref nums[leftIdx], ref nums[rightIdx]);
                rightIdx--;
            }
            else{
                leftIdx++;
            }
        }
        
        return nums;
    }
}